taichi/transforms/flag_access.cpp

```diff
@@ -110,16 +121,13 @@ class WeakenAccess : public BasicStmtVisitor {
       if (is_struct_for) {
         bool same_as_loop_snode = true;
         for (auto snode : stmt->snodes.data) {
-          if (snode->type == SNodeType::place) {
-            snode = snode->parent;
-          }
           SNode *loop_snode = nullptr;
           if (current_struct_for) {
             loop_snode = current_struct_for->snode;
           } else {
             loop_snode = current_offload->snode;
           }
-          if (snode != loop_snode) {
+          if (!share_sparsity(snode, loop_snode)) {
             same_as_loop_snode = false;
           }
           TI_ASSERT(loop_snode);
```

Comment:
Let's move this line above `share_sparsity(snode, loop_snode)` because that function will crash if `loop_snode` is `nullptr`.

BTW, shall we add something like `if (!same_as_loop_snode) break;` here to reject early?

Reply:
> Let's move this line above `share_sparsity(snode, loop_snode)` because that function will crash if `loop_snode` is `nullptr`.

Good point!



> BTW, shall we add something like `if (!same_as_loop_snode) break;` here to reject early?

I thought about that, but here we have many loops and I'd rather keep the control flow easy to understand. This part is not the bottleneck anyway :-)